crates/wasmparser/src/readers/core/operators.rs

```diff
@@ -1176,3 +1187,1280 @@ impl<'a> Iterator for OperatorsIteratorWithOffsets<'a> {
+            Operator::F64x2Sqrt => self.visit_f64x2_sqrt(input),
+            Operator::F64x2Add => self.visit_f64x2_add(input),
+            Operator::F64x2Sub => self.visit_f64x2_sub(input),
+            Operator::F64x2Mul => self.visit_f64x2_mul(input),
+            Operator::F64x2Div => self.visit_f64x2_div(input),
+            Operator::F64x2Min => self.visit_f64x2_min(input),
+            Operator::F64x2Max => self.visit_f64x2_max(input),
+            Operator::F64x2PMin => self.visit_f64x2_p_min(input),
+            Operator::F64x2PMax => self.visit_f64x2_p_max(input),
+            Operator::I32x4TruncSatF32x4S => self.visit_i32x4_trunc_sat_f32x4_s(input),
+            Operator::I32x4TruncSatF32x4U => self.visit_i32x4_trunc_sat_f32x4_u(input),
+            Operator::F32x4ConvertI32x4S => self.visit_f32x4_convert_i32x4_s(input),
+            Operator::F32x4ConvertI32x4U => self.visit_f32x4_convert_i32x4_u(input),
+            Operator::I32x4TruncSatF64x2SZero => self.visit_i32x4_trunc_sat_f64x2_s_zero(input),
+            Operator::I32x4TruncSatF64x2UZero => self.visit_i32x4_trunc_sat_f64x2_u_zero(input),
+            Operator::F64x2ConvertLowI32x4S => self.visit_f64x2_convert_low_i32x4_s(input),
+            Operator::F64x2ConvertLowI32x4U => self.visit_f64x2_convert_low_i32x4_u(input),
+            Operator::F32x4DemoteF64x2Zero => self.visit_f32x4_demote_f64x2_zero(input),
+            Operator::F64x2PromoteLowF32x4 => self.visit_f64x2_promote_low_f32x4(input),
+            Operator::I8x16RelaxedSwizzle => self.visit_i8x16_relaxed_swizzle(input),
+            Operator::I32x4RelaxedTruncSatF32x4S => self.visit_i32x4_relaxed_trunc_sat_f32x4_s(input),
+            Operator::I32x4RelaxedTruncSatF32x4U => self.visit_i32x4_relaxed_trunc_sat_f32x4_u(input),
+            Operator::I32x4RelaxedTruncSatF64x2SZero => self.visit_i32x4_relaxed_trunc_sat_f64x2_s_zero(input),
+            Operator::I32x4RelaxedTruncSatF64x2UZero => self.visit_i32x4_relaxed_trunc_sat_f64x2_u_zero(input),
+            Operator::F32x4Fma => self.visit_f32x4_fma(input),
+            Operator::F32x4Fms => self.visit_f32x4_fms(input),
+            Operator::F64x2Fma => self.visit_f64x2_fma(input),
+            Operator::F64x2Fms => self.visit_f64x2_fms(input),
+            Operator::I8x16LaneSelect => self.visit_i8x16_lane_select(input),
+            Operator::I16x8LaneSelect => self.visit_i16x8_lane_select(input),
+            Operator::I32x4LaneSelect => self.visit_i32x4_lane_select(input),
+            Operator::I64x2LaneSelect => self.visit_i64x2_lane_select(input),
+            Operator::F32x4RelaxedMin => self.visit_f32x4_relaxed_min(input),
+            Operator::F32x4RelaxedMax => self.visit_f32x4_relaxed_max(input),
+            Operator::F64x2RelaxedMin => self.visit_f64x2_relaxed_min(input),
+            Operator::F64x2RelaxedMax => self.visit_f64x2_relaxed_max(input),
+        }
+    }
+
+    fn visit_nop(&mut self, input: Input) -> Self::Output;
```

Comment:
Personally I still think that this should be `Result<Self::Output>` since this primarily intended for the validator right now which already returns results (and otherwise avoids `foo??`). You mentioned there's a perf difference, though, so is there something I can look at commit-wise to help drill down?

Reply:
Here is were our misunderstanding comes from: This trait and API is not just designed for the validator. If we want to replace both unnecessary unpredictable branches we have the need to design this trait for user side as well since the 2nd unnecessary unpredictable branch is in user code. So the `VisitOperator` trait is expected to be implemented by users of the `wasmparser` trait.

For example: In `wasmi` we want to implement `VisitOperator` by the component that translates from Wasm bytecode to our internal `wasmi` bytecode. The very same component will also drive the validation via this trait.